### ISSUE TITLE
DASH streams stall unpredictably after manifest switch from dynamic to static

### ISSUE THREAD

**Issue by martinstark** [shaka-project/shaka-player#3902]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes, there are a number of issues peripherally associated to this one, which all point back towards b3166db. 

**What version of Shaka Player are you using?**
3.3.0 (reproduced on most versions between 3.1.2 - 3.3.0). 

We have been unable to reproduce the issue on 3.1.1 and have stayed on that version until recently.

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `master`?**
Yes

**Are you using the demo app or your own custom app?**
Custom

**If custom app, can you reproduce the issue using our demo app?**
N/A

**What browser and OS are you using?**
Windows / Mac / Linux, Chrome / Firefox

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
N/A

**What are the manifest and license server URIs?**

Due to the dynamic nature of the issue, reproducing it is quite involved, time sensitive and time consuming. I can help arranging an environment where it's possible to reproduce, should it be needed.

**What configuration are you using?  What is the output of `player.getConfiguration()`?**

<details>
  <summary>COnfig</summary>
{
  "drm": {
    "retryParameters": {
      "maxAttempts": 5,
      "baseDelay": 1200,
      "backoffFactor": 2.2,
      "fuzzFactor": 0.4,
      "timeout": 40000,
      "stallTimeout": 18000,
      "connectionTimeout": 6000
    },
    "servers": {
    },
    "clearKeys": {},
    "advanced": {},
    "delayLicenseRequestUntilPlayed": false,
    "logLicenseExchange": false,
    "updateExpirationTime": 1,
    "preferredKeySystems": []
  },
  "manifest": {
    "retryParameters": {
      "maxAttempts": 5,
      "baseDelay": 1500,
      "backoffFactor": 2.2,
      "fuzzFactor": 0.4,
      "timeout": 40000,
      "stallTimeout": 18000,
      "connectionTimeout": 12000
    },
    "availabilityWindowOverride": null,
    "disableAudio": false,
    "disableVideo": false,
    "disableText": false,
    "disableThumbnails": false,
    "defaultPresentationDelay": 20,
    "dash": {
      "clockSyncUri": "https://time.akamai.com/?iso",
      "ignoreDrmInfo": false,
      "disableXlinkProcessing": false,
      "xlinkFailGracefully": false,
      "ignoreMinBufferTime": true,
      "autoCorrectDrift": true,
      "initialSegmentLimit": 1000,
      "ignoreSuggestedPresentationDelay": false,
      "ignoreEmptyAdaptationSet": false,
      "ignoreMaxSegmentDuration": false,
      "keySystemsByURI": {
      }
    },
    "hls": {
      "ignoreTextStreamFailures": false,
      "ignoreImageStreamFailures": false,
      "useFullSegmentsForStartTime": true,
      "defaultAudioCodec": "mp4a.40.2",
      "defaultVideoCodec": "avc1.42E01E"
    }
  },
  "streaming": {
    "retryParameters": {
      "maxAttempts": 6,
      "baseDelay": 1500,
      "backoffFactor": 2.5,
      "fuzzFactor": 0.4,
      "timeout": 40000,
      "stallTimeout": 18000,
      "connectionTimeout": 12000
    },
    "rebufferingGoal": 4,
    "bufferingGoal": 60,
    "bufferBehind": 30,
    "ignoreTextStreamFailures": true,
    "alwaysStreamText": false,
    "startAtSegmentBoundary": false,
    "gapDetectionThreshold": 0.1,
    "smallGapLimit": 0.5,
    "jumpLargeGaps": true,
    "durationBackoff": 1,
    "forceTransmuxTS": false,
    "safeSeekOffset": 5,
    "stallEnabled": true,
    "stallThreshold": 1,
    "stallSkip": 0.1,
    "useNativeHlsOnSafari": true,
    "inaccurateManifestTolerance": 2,
    "lowLatencyMode": false,
    "autoLowLatencyMode": false,
    "forceHTTPS": false,
    "preferNativeHls": false,
    "updateIntervalSeconds": 1,
    "dispatchAllEmsgBoxes": false,
    "observeQualityChanges": false
  },
  "offline": {
    "usePersistentLicense": true
  },
  "abr": {
    "enabled": true,
    "useNetworkInformation": true,
    "defaultBandwidthEstimate": 1000000,
    "switchInterval": 3,
    "bandwidthUpgradeTarget": 0.75,
    "bandwidthDowngradeTarget": 0.85,
    "restrictions": {
      "minWidth": 0,
      "maxWidth": null,
      "minHeight": 0,
      "maxHeight": null,
      "minPixels": 0,
      "maxPixels": null,
      "minFrameRate": 0,
      "maxFrameRate": null,
      "minBandwidth": 0,
      "maxBandwidth": null
    },
    "advanced": {
      "minTotalBytes": 128000,
      "minBytes": 16000,
      "fastHalfLife": 2,
      "slowHalfLife": 5
    }
  },
  "preferredAudioLanguage": "sv",
  "preferredTextLanguage": "",
  "preferredVariantRole": "",
  "preferredTextRole": "",
  "preferredAudioChannelCount": 2,
  "preferredVideoCodecs": [],
  "preferredAudioCodecs": [],
  "preferForcedSubs": false,
  "preferredDecodingAttributes": [],
  "restrictions": {
    "minWidth": 0,
    "maxWidth": null,
    "minHeight": 0,
    "maxHeight": null,
    "minPixels": 0,
    "maxPixels": null,
    "minFrameRate": 0,
    "maxFrameRate": null,
    "minBandwidth": 0,
    "maxBandwidth": null
  },
  "playRangeStart": 0,
  "playRangeEnd": null,
  "cmcd": {
    "enabled": false,
    "sessionId": "",
    "contentId": "",
    "useHeaders": false
  }
}
</details>

**What did you do?**

Play a "startover" dynamic live stream that switches to static

**What did you expect to happen?**
The stream continues playing until the end is reached

**What actually happened?**

The stream **sometimes** randomly freezes with an endless spinner. It does not always happen. Due to the time consuming nature of reproducing the issue (can range from a few minutes to a few hours to reproduce it), we have a limited number of observations:

1. me and a colleague started the same stream with the same manifest at roughly the same time, one of the streams froze and the other did not
2. me and a colleague started the same stream 86 seconds apart, our streams froze 84 seconds apart
3. it seems to happen about 1/4 or 1/5 of attempts

The freeze is due to Shaka being unable to find a segment reference, causing shaka to stall in an infinite loop of being unable to fill the buffer. The symptoms are slightly different almost every time it happens, making it hard to track down a root cause.

Most of these screenshots are taken from different testing sessions, with different amounts of logging enabled. I have labelled each of them for easier referencing. Most of the logs are custom, since turning on verbose shaka logging for a longer period causes the browser console to be unusable over longer sessions.

The majority of sessions ended up freezing in the "Something is buffered from another Stream." part of `getSegmentReferenceNeeded_`, which then logs a neat warning in debug mode. This coincided with a switch that cleared the current `segmentIterator` of the video stream, but not the audio stream. At this point I thought I was on track to find the root cause, but the latest freeze threw me.

### Session 1

This was the first time we started getting a clue as to what caused shaka to stumble.

![image](https://user-images.githubusercontent.com/901824/150978639-3b3ffa08-31b2-4db6-96e0-364fdcebbc24.png)

### Session 2

Reproducing the issue again with more logging, it is clearly visible that Shaka has references to all segments in the stream, but stalls due to being able to find the next segment reference (ref is null).

Stall:

![image](https://user-images.githubusercontent.com/901824/150979608-10797a81-26ed-439d-8912-16d28610618b.png)

References array has segments up until `duration` / `appendWindowEnd`.

![image](https://user-images.githubusercontent.com/901824/150979624-80fd6ce8-6aab-48c5-b100-e90bfa32119e.png)

The segment shaka is unable to find (at time 2457.6) exists in the references array, as it looked after the manifest switched to static:

![image](https://user-images.githubusercontent.com/901824/150979848-c9a5e344-91c2-4d60-b228-08a342339129.png)

### Session 3

Just another screenshot, not much new information.

![image](https://user-images.githubusercontent.com/901824/150980071-49839cf9-254f-40f1-be05-4c00291bfc94.png)

### Session 4

This session was interesting in that it froze using the active `segmentIterator` ("Something is buffered from the same Stream" part of
`getSegmentReferenceNeeded_`), and the `appendWindowEnd` differing between the audio and video tracks after the manifest turned static.

These were the final two segments shaka could find before video started returning `null` references, this time _without_ the little yellow warning since it returns `mediaState.segmentIterator.current()` without checking if it's null or doing any logging around it.

Notably, `appendWindowEnd` on the video is set to `Infinity`, while audio has the correct end time.

![image](https://user-images.githubusercontent.com/901824/150981244-aa75b695-6ec2-4012-9e74-5785db82f1fe.png)

The next attempt to get a video reference:

![image](https://user-images.githubusercontent.com/901824/150981814-1b6b425a-8198-4f00-9e30-2f18cd4b2405.png)

I will continue to debug this, but at this point I'd be very happy if someone with more knowledge of the shaka source could point me in the right direction.

**Comment by martinstark** [comment]
This is still an issue in 3.3.1.

We've downgraded to 3.1.1 again to get around the stalling issue for now.

**Comment by martinstark** [comment]
Related: https://github.com/shaka-project/shaka-player/issues/2581

To enable playing the dynamic -> static manifests in shaka we use a network filter.

The network filter modifies the manifest before it is fed to shaka, by taking the `t` value of the `SegmentTimeline` `S` and applying it as the `presentationTimeOffset` on the `SegmentTemplate` **while the manifest is still dynamic**.

![image](https://user-images.githubusercontent.com/901824/161540699-4dc45ef6-589e-4ce8-8b6d-197be6690668.png)

This has worked like a charm on versions of shaka up to and including 3.1.1. The commit that made it stop working is https://github.com/shaka-project/shaka-player/commit/b3166db4bc7cac968a03796bc9c2d7e598fb88c7. Since this commit was made, the streams have started stalling unpredictably after the manifest turns static, with shaka unable to find the correct segment in maybe 20% of stream starts.

![image](https://user-images.githubusercontent.com/901824/161540606-4970ee5f-23b2-403a-8188-a941ca6a5df8.png)

Internally, Shaka holds references to full arrays of variant segments, and shows the correct seekRange/duration, but can not play past a certain point in the stream due to some internal logic going wrong that I haven't been able to pin down.

I was able to reproduce the issue using the reference stream from Unified Streaming:

```
echo "https://demo.unified-streaming.com/k8s/live/stable/scte35.isml/.mpd?vbegin=$(($(date -u +%s)-2000))&vend=$(($(date -u +%s)+30))"
```
- apply network filter
- start stream, pause and wait for manifest to turn static
- attempt seeking close to end and let play
- stream may or may not stall, with shaka unable to internally resolve the next segment

**Comment by martinstark** [comment]
The custom network filter is no longer needed to play the type of DASH manifest listed in this issue, and it also seems like the issue has been unintentionally fixed for DASH in Shaka [4.1.2](https://github.com/shaka-project/shaka-player/releases/tag/v4.1.2). I can see there are a number of changes to presentation timeline and segment calculations in the release, but haven't dug into what the change that created the fix might be.

There is still a (newly detected) stalling issue on HLS.

**Comment by joeyparrish** [comment]
The most likely change to affect this in v4.1.2 is "Fix segment index assertions with DAI ([2ef9270](https://github.com/shaka-project/shaka-player/commit/2ef9270a95dd0de5119dc5c6fe6cc18a59202c24))".

May we close this issue as fixed now?

**Comment by martinstark** [comment]
I'll close it and re-open if something crops up again. It does look like DASH is working, but we're now running into an identical stall issue on HLS instead. Since native Fairplay playback is broken in 4.1.2 we've been testing out setting `useNativeHlsOnSafari: false`, which made the issue show up. It probable deserves its own issue if we decide to spend more time on debugging it.
